### ISSUE TITLE
microdom shouldn't bork on malformed tags when beExtremelyLenient is enabled

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2225]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#2225|
|Type|enhancement|
|Created|2006-11-08 22:07:51Z|

When microdom encounters things like `<div al!\n ign="center">`, it should ignore the whole tag if `beExtremelyLenient` is enabled.

Justification: Firefox, Safari and IE all ignore malformed tags, and the purpose of beExtremelyLenient is to let us get a handle on bad HTML.

<details><summary>Searchable metadata</summary>

```
trac-id__2225 2225
type__enhancement enhancement
reporter__jml jml
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1163023671000000 1163023671000000
changetime__1163137278000000 1163137278000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml set owner to @glyph|
|-|-|

Ready for review in `microdom-bad-tags-2225`

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph set owner to @jml<br>@glyph set status to `new`|
|-|-|

I basically like this.

I'd rather not have whitespace mashing in files which aren't otherwise modified (I don't see any semantic changes in microdom.py, and the tests pass if I revert it), but this has to happen some time, so now that I've gone to the trouble of looking at it let's go ahead and merge it.

The only problem I have is that I don't think the string "True" has any particular semantics in this case.  In some cases, XHTML requires a string like "on" or "checked"; in others it accepts any value (including the empty string).

If "True" is random here, I'd suggest replacing it with the empty string, then merge.  Otherwise go ahead and merge as-is.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

Guh.  Trac really freaked out Firefox somehow and messed up a bunch of attributes as part of that review.  Puttin gthem back...

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_5"></a>|@jml commented|
|-|-|

```
[16:32]  <jml> glyph: so, sux uses "True" as the default value for an attribute.
[16:33]  <jml> glyph: you want me to change that (totally untested) behaviour?
[16:33]  <glyph> jml: Nope
[16:33]  <glyph> jml: that's a good enough reason to leave as-is
[16:33]  <glyph> jml: perhaps we should revisit it later
[16:33]  <jml> sure
```

[#2227](https://github.com/twisted/twisted/issues/2227) is the revisit-it-later ticket.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_6"></a>|@jml set status to `closed`|
|-|-|

(In [18633]) Make sux even more lenient when `beExtremelyLenient` is enabled.
 * Author: jml
 * Reviewer: glyph
 * Fixes [#2225](https://github.com/twisted/twisted/issues/2225)

Don't explode when we find weird characters and line breaks in HTML. 
Instead, muddle on as best we can. This behaviour inspired by the way
Firefox handles really bad HTML.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
